src/chapter5/message-passing.md

```diff
@@ -1 +1,240 @@
 # Message Passing
+
+## Primary MPI Routines
+
+``` C
+int MPI_Init(int * argc, char ** argv);
+// initializes the MPI environment.
+//Argc argv are the parameters come
+//from main(argc,argv). The return value is an
+// error code. 0 is OK. Non-zero is an error code
+```
+
+``` C
+int MPI_Comm_size(MPI_Comm comm, int \* size);
+// this functions gets the number of MPI processes
+// i.e. the number you enter when you go mpirun -np \<size> myprog.exe
+// *size is C syntax indicating that size will be modified to contain
+// the value after the function returns. The return value is only used
+// for error detection. printf(“MPI size is %d\n”,size);
+int MPI_Comm_rank(MPI_Comm comm, int \* rank);
+// this returns the rank of this particular process
+// rank contains the value for that process- the function return value is an error code
+```
+
+![MPI routines](../imgs/mpi_routines.png)
+
+### Point-to-Point communication
+
+These are blocking functions - they wait until the message is sent or received. Note that the CPU is actively polling the network interface when waiting for a message. This is opposite in behaviour to other C functions, i.e. c= getChar() (which causes a context switch and then a sleep in the OS). This is done for speed reasons.
+
+```C
+int MPI_Send(void * buf, int count, MPI_Datatype type, int dest, int tag, MPI_Comm comm);
+```
+
+Sends a message from the calling process to another process
+
+INPUT PARAMETERS
+
+- buf
+  - Initial address of send buffer (choice).
+- count
+  - Number of elements sent (non negative integer).
+- type
+  - DataType of each send buffer element (handle).
+- dest
+  - Rank of destination (integer).
+- tag
+  - Message tag (integer).
+- comm
+  - Communicator (handle).
+
+OUTPUT PARAMETER
+
+- IERROR
+  - Fortran only: Error status (integer).
+
+```c
+int MPI_Recv(void * buf, int count, MPI_Datatype type, int source, int tag, MPI_Comm comm, MPI_Status * status);
+```
+
+Receives a message from another process
+
+INPUT PARAMETERS
+
+- count
+  - Maximum number of elements to receive (integer).
+- type
+  - DataType of each receive buffer entry (handle).
+- source
+  - Rank of source (integer).
+- tag
+  - Message tag (integer).
+- comm
+  - Communicator (handle).
+
+OUTPUT PARAMETERS
+
+- buf
+  - Initial address of receive buffer (choice).
+- status
+  - Status object (status).
+- IERROR
+  - Fortran only: Error status (integer).
+
+### Primary MPI Routines closing
+
+In a header file you will find
+
+``` C
+int MPI_Finalize(void);
+```
+
+To call in your C/C++ program
```

Comment:
Pedantic not but 'C/C++' should be 'C or C++'. The notation 'C/C++' often gets confused as if the language is 'C/C++' or the usual or meaning that / is used for.